Maple2.Server.Game/Manager/ItemBoxManager.cs

```diff
@@ -0,0 +1,284 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Maple2.Database.Storage;
+using Maple2.Model;
+using Maple2.Model.Enum;
+using Maple2.Model.Error;
+using Maple2.Model.Game;
+using Maple2.Model.Metadata;
+using Maple2.Server.Game.Session;
+
+namespace Maple2.Server.Game.Manager;
+
+public class ItemBoxManager {
+    #region Autofac Autowired
+    // ReSharper disable MemberCanBePrivate.Global
+    public required ItemMetadataStorage ItemMetadata { private get; init; }
+    public required TableMetadataStorage TableMetadata { private get; init; }
+    // ReSharper restore All
+    #endregion
+
+    private readonly GameSession session;
+    public int BoxCount;
+
+    public ItemBoxManager(GameSession session) {
+        this.session = session;
+        BoxCount = 0;
+    }
+
+    public void Reset() {
+        BoxCount = 0;
+    }
+
+    public ItemBoxError Open(Item item, int count = 1, int index = 0) {
+        int[] itemBoxParams = item.Metadata.Function?.Parameters.Split(',').Select(int.Parse).ToArray() ?? Array.Empty<int>();
+        return item.Metadata.Function?.Type switch {
+            ItemFunction.SelectItemBox => SelectItemBox(item, itemBoxParams, index, count),
+            ItemFunction.OpenItemBox => OpenItemBox(item, itemBoxParams, count),
+            ItemFunction.OpenItemBoxWithKey => OpenItemBoxWithKey(item, itemBoxParams, count),
+            _ => throw new ArgumentOutOfRangeException(item.Metadata.Function?.Type.ToString(), "Invalid box type"),
+        };
+    }
+
+    private ItemBoxError SelectItemBox(Item item, int[] itemBoxParams, int index, int count = 1) {
+        var error = ItemBoxError.ok;
+
+        if (itemBoxParams.Length < 2) {
+            return ItemBoxError.s_err_cannot_open_multi_itembox_inventory_fail;
+        }
+
+        // check if box count is enough to open
+        if (session.Item.Inventory.Find(item.Id).Sum(box => box.Amount) < itemBoxParams[0]) {
+            return ItemBoxError.s_err_cannot_open_multi_itembox_inventory_fail;
+        }
+
+        // check if we can get metadata
+        if (!TableMetadata.IndividualItemDropTable.Entries.TryGetValue(itemBoxParams[1], out Dictionary<byte, IList<IndividualItemDropTable.Entry>>? dropGroupTable)) {
+            return ItemBoxError.s_err_cannot_open_multi_itembox_inventory_fail;
+        }
+
+        // SelectItemBox disregards DropGroup
+        IList<IndividualItemDropTable.Entry> drops = session.ItemBox.FilterDrops(dropGroupTable.Values.SelectMany(x => x).ToList());
+
+        if (drops.Count == 0) {
+            return ItemBoxError.s_err_cannot_open_multi_itembox_inventory_fail;
+        }
+
+        IndividualItemDropTable.Entry selectedEntry = drops[index];
+        for (int startCount = 0; startCount < count; startCount++) {
+            if (!session.Item.Inventory.ConsumeItemComponents(new List<ItemComponent>() {
```

Comment:
Here and elsewhere, prefer array instead of List if you are don't need to add to it afterwards.

```cs
new []{new ItemComponent(item.Id, -1, itemBoxParams[0], ItemTag.None)}
```